### ISSUE TITLE
~7% performance drop in `box_index_get()`

### ISSUE THREAD

**Issue by Gumix** [tarantool/tarantool#9211]
Exposed by the `TreeGetOneRandomExistingKey` benchmark (it is not in the master yet):
- https://github.com/tarantool/tarantool/pull/7083

```
Tarantool 3.0.0-alpha3-57-g8e4f263e9
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=TRUE
Compiler: GNU-12.2.0
C_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/home/iverbin/git/tarantool=. -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type
CXX_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/home/iverbin/git/tarantool=. -std=c++11 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type
```

**How to run:**

```
$ taskset 0x2 ./build/perf/memtx.perftest \
  --benchmark_min_warmup_time=5 \
  --benchmark_repetitions=30 \
  --benchmark_report_aggregates_only=true \
  --benchmark_display_aggregates_only=true \
  --benchmark_filter=TreeGetOneRandomExistingKey
```

```
----------------------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------
MemtxFixture/TreeGetOneRandomExistingKey_mean          382 ns          382 ns           30 items_per_second=2.62324M/s
MemtxFixture/TreeGetOneRandomExistingKey_median        384 ns          384 ns           30 items_per_second=2.60215M/s
MemtxFixture/TreeGetOneRandomExistingKey_stddev       17.1 ns         17.1 ns           30 items_per_second=115.554k/s
MemtxFixture/TreeGetOneRandomExistingKey_cv           4.47 %          4.47 %            30 items_per_second=4.41%
```

The guilty commit is `box: introduce template parameters for the sort order check` from:
- https://github.com/tarantool/tarantool/pull/8915

Before that commit:
```
----------------------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------
MemtxFixture/TreeGetOneRandomExistingKey_mean          354 ns          354 ns           30 items_per_second=2.83292M/s
MemtxFixture/TreeGetOneRandomExistingKey_median        350 ns          350 ns           30 items_per_second=2.85371M/s
MemtxFixture/TreeGetOneRandomExistingKey_stddev       16.2 ns         16.2 ns           30 items_per_second=128.169k/s
MemtxFixture/TreeGetOneRandomExistingKey_cv           4.58 %          4.58 %            30 items_per_second=4.52%
```

It's not clear what caused the regression. According to `perf diff`, before the commit `mp_next` and `mp_compare_uint` were inlined to `TupleCompareWithKey<...>::compare`, while after the commit they aren't. Maybe this is the reason for the performance degradation, or maybe not:

```
$ perf diff ./before/perf.data ./after/perf.data
# Event 'cpu-clock:pppH'
#
# Baseline  Delta Abs  Shared Object         Symbol                                                                                                                                   >
# ........  .........  ....................  .........................................................................................................................................>
#
    45.88%    -21.32%  memtx.perftest        [.] (anonymous namespace)::TupleCompareWithKey<0, 0, 1, 1, 1, 2, 1>::compare
              +13.35%  memtx.perftest        [.] mp_next
              +10.71%  memtx.perftest        [.] mp_compare_uint
    18.61%     -1.83%  memtx.perftest        [.] memtx_tree_index_get_internal<false>
     1.39%     +0.79%  memtx.perftest        [.] space_by_id
     1.23%     -0.52%  memtx.perftest        [.] txn_begin_ro_stmt
     1.48%     -0.49%  memtx.perftest        [.] access_check_space
```

**Comment by CuriousGeorgiy** [comment]
2 cents from me: actually, if we look at the CV[^1], which is around 4%, the actual performance degradation is around `7% - 4% =3%`. But nevertheless it's cool that you've identified the possible cause.

[^1]: https://en.wikipedia.org/wiki/Coefficient_of_variation

**Comment by mkostoevr** [comment]
Reproduced similar results for gcc in `tuple_tuple_compare` and `tuple_tuple_compare_hint` benchmarks, the problem is difference between gcc's and clang's inline thresholds (clang has one expected regression fixed with the last commit, whereas gcc has two noticable ones: on the [comparator refactoring](https://github.com/tarantool/tarantool/commit/1609b34c23a5a82dd6d10db4e7ad9c716fe6659b) and on the [sort order templatification](https://github.com/tarantool/tarantool/commit/9cc5a0bb1b3fdc155e6bec35e84699aff75a56f6)).

The first one for gcc can be fixed by marking the `key_part_compare_fields` as `inline`, the second one can be decreased by marking the `key_part_hint` as `inline`. With this configuration the `tuple.perftest` benchmark's gcc regression is mitigated and the clang performance is not affected.

Benchmarks (`RelWithDebInfo` configuration):

| Version | `tuple_tuple_compare` | `tuple_tuple_compare_hint` |
| - | - | - |
| clang before #8915 | 17.2 ns | 19.9 ns |
| clang after #8915 | 16.9 ns | 20.1 ns |
| gcc before #8915 | 19.0 ns | 24.5 ns |
| gcc after #8915 | 22.8 ns | 23.2 ns |
| gcc after #8915 with suggested changes | 19.2 ns | 23.1 ns |

**Comment by mkostoevr** [comment]
The probmlem is that GCC has a limit for inliner: it refuses to inline a function is in the result the estimated amount of pseudoinstructions in the compilation unit will exceed 30% (by default). The PR#9248 increases the limit to 100% and so mitigates the regression, but it's not fixed fully because of sife effects of this option.

It's desided to freeze this issue until we have an establihed automatical benchmarking system.

**Comment by CuriousGeorgiy** [comment]
@mkostoevr why can't you simply use `__attribute__((always_inline))`?

**Comment by mkostoevr** [comment]
@CuriousGeorgiy one can't simply use [__attribute__((always_inline))](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-always_005finline-function-attribute) to relyably improve code performance in this case.

Now it shows the same results on Tigerlake laptop, but last time I measured it on Zen 3 desktop it showed a regression in `tuple.perftest`.

**Comment by mkostoevr** [comment]
@CuriousGeorgiy my bad, measured wrong binaries, here're execution time changes over the #9248.

TreeGetRandomExistingKeys: $\color{green}-2.6％$
TreeGet1RandomExistingKey: $\color{green}-2％$
tuple_tuple_compare: $\color{RedOrange}+1.3％$
tuple_tuple_compare_hint: $\color{red}+15.4％$

This slows tuple.perftest result down on Tigerlake laptop too.

These optimizations are very controversal so I would suggest to focus on higher level instead of tweaking source and commandline to satisfy compilers.

**Comment by CuriousGeorgiy** [comment]
@mkostoevr I agree, let's move on.

Can you please post the degradation we have left after applying #9220 and close the ticket?

BTW, what compiler version are you using (please don't forget to include this in future performance reports)? Did you try using `gcc-13`? Problems may as well go away on fresh compiler versions.

**Comment by mkostoevr** [comment]
Need to draw a context. We have two scopes here.

### tuple.perftest

The sort order implementation (#8915) introduced the following changes in the `tuple.perftest` with GCC compiler:

 | Test | Tigerlake | Zen 3 |
| - | - | - |
| tuple_tuple_compare | $\color{red}+2.5％$ | $\color{red}+20.9％$ |
| tuple_tuple_compare_hint | $\color{green}-22.1％$ | $\color{green}-5.4％$ |

The first patch (#9220) fixed the Tigerlake regression and mitigated the Zen 3 one. So after merging it we have the following results (compared with performance prior to #8915):

 | Test | Tigerlake | Zen 3 |
| - | - | - |
| tuple_tuple_compare | $\color{green}-2.7％$ | $\color{red}+2.6％$ |
| tuple_tuple_compare_hint | $\color{green}-16.5％$ | $\color{green}-3.1％$ |

Zen 3-specific regression root cause is under investigation.

### memtx.perftest

The sort order implementation (#8915) introduced the following changes in the `memtx.perftest` with GCC compiler:

| Test | Tigerlake | Zen 3 |
| - | - | - 
| TreeGetRandomExistingKeys | $\color{red}+3％$ | Coming soon... |
| TreeGet1RandomExistingKey | $\color{red}+14.3％$ | Coming soon... |

The corresponding patch (#9248) mitigates the regression (though, does not fix it fully), and new results are the wollowing (compared with performance prior to #8915):

| Test | Tigerlake | Zen 3 |
| - | - | - |
| TreeGetRandomExistingKeys | $\color{RedOrange}+0.6％$ | Coming soon... |
| TreeGet1RandomExistingKey | $\color{RedOrange}+2％$ | Coming soon... |

Since the PR fixes the issue by adding a compiler commandline argument, it's desided to discontinue it until we have a strong performance measurement system established, because this can lead to unexpected side-effects in the future.

### TLDR

The degradation we have now:
- `perf/tuple.cc::tuple_tuple_compare`: 2.6% on Zen 3, GCC build.
- `perf/memtx.cc::TreeGetRandomExistingKeys`: 3% on GCC build if [#9248](https://github.com/tarantool/tarantool/pull/9248) is not merged, else 0.6%.
- `perf/memtx.cc::TreeGet1RandomExistingKey`: 14.3% on GCC build if [#9248](https://github.com/tarantool/tarantool/pull/9248) is not merged, else 2%.

No degradation at all with Clang.
